operator/v1/types.go

```diff
@@ -52,6 +52,33 @@ type ClusterManagerSpec struct {
 	// +optional
 	// +kubebuilder:default={mode: Default}
 	DeployOption DeployOption `json:"deployOption,omitempty"`
+
+	// Detached include configurations we needs in the detached mode.
+	// +optional
+	Detached Detached `json:"detached"`
+}
+
+// Detached represents customized configurations we need to set in the detached mode.
+type Detached struct {
+	// RegistrationAPIServiceTargetPort is used in rendering the file "manifests/cluster-manager/hub/cluster-manager-registration-webhook-apiservice.yaml".
+	//
+	// +optional
+	RegistrationWebhookAPIService APIService `json:"registrationWebhookAPIService"`
+
+	// WorkAPIWebhookAPISerivce is used in redering the file "manifests/cluster-manager/hub/cluster-manager-work-webhook-apiservice.yaml".
+	//
+	// +optional
+	WorkWebhookAPIService APIService `json:"workWebhookAPIService"`
+}
+
+// APIService represents customized configurations we need to set for APIService.
+type APIService struct {
+	// ServicePort represents the `Spec.Service.Port` in APIService.
+	// If this field is not set or equals to 0, we will use the default value: 443
+	//
+	// +optional
+	// +kubebuilder:default=443
+	ServicePort int32 `json:"servicePort"`
```

Comment:
why only port, I think you would still need address and cabundle

Reply:
For the ca-bundle, based on the current implementation, this filed is using the `ca-bundle-configmap` generated by certrotation-controller.

For the address, the strategy is as the following:

This is the manifest file we need to render:
```
apiVersion: apiregistration.k8s.io/v1
kind: APIService
metadata:
  name: v1.admission.cluster.open-cluster-management.io
spec:
  group: admission.cluster.open-cluster-management.io
  version: v1
  service:
    name: cluster-manager-registration-webhook
    namespace: {{ .ClusterManagerNamespace }}
    port: {{.RegistrationServicePort}} # This is what we change
  caBundle: {{ .RegistrationAPIServiceCABundle }}
  groupPriorityMinimum: 10000
  versionPriority: 20
```
We only change the `port` field and apiservice always point to the service -- `cluster-manager. cluster-manager-registration-webhook`.
However, in the detached mode, this service needs users to create by themselves. ExternalName, InCluster, LoadBalancer and so on, they create whatever they need just with the fixed service name.

Reply:
If a user need to set such service, why do we still need to set the port. The user can configure the port to what it needs, right?

Reply:
It's true, but as my experience of tests, it's not handy. 

I tested last night as the following.

![image](https://user-images.githubusercontent.com/11444421/150480397-6f8a4d20-295e-492d-a0af-874c5ff07ac5.png)

And my configuration of Ingress would be:

```yaml
cat <<EOF | kubectl apply -f -
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: cluster-manager-webhook-ingress
  namespace: cluster-manager
spec:
  rules:
  - http:
      paths:
      - pathType: Prefix
        path: "/apis/admission.cluster.open-cluster-management.io/v1"
        backend:
          service:
            name: cluster-manager-registration-webhook
            port:
              number: 443
      - pathType: Prefix
        path: "/apis/admission.work.open-cluster-management.io/v1"
        backend:
          service:
            name: cluster-manager-work-webhook
            port:
              number: 443
```

I see the logs of ingress-controller shows the traffic is direct to the right pod, but the response is 400.

Then I guess the ingress-controller is not transparent enough, and then I tried many different nginx parameters but all failed. 😭

After hours of testing, finally I decided to make service ports of APISerivce configurable to leverage NodePort.  In other word, sacrificing simplicity and consistency to gain flexibility.